src/art/art.c

```diff
@@ -1448,118 +1680,327 @@ static void art_node_printf(const art_node_t *node, uint8_t depth) {
+            }
+        } break;
+        case CROARING_ART_NODE16_TYPE: {
+            art_node16_t *node16 = (art_node16_t *)art_deref(art, ref);
+            for (uint8_t i = 0; i < node16->count; ++i) {
+                node16->children[i] =
+                    art_move_node_to_shrink(art, node16->children[i]);
+                art_shrink_at(art, node16->children[i]);
+            }
+        } break;
+        case CROARING_ART_NODE48_TYPE: {
+            art_node48_t *node48 = (art_node48_t *)art_deref(art, ref);
+            for (int i = 0; i < 256; ++i) {
+                if (node48->keys[i] != CROARING_ART_NODE48_EMPTY_VAL) {
+                    uint8_t idx = node48->keys[i];
+                    node48->children[idx] =
+                        art_move_node_to_shrink(art, node48->children[idx]);
+                    art_shrink_at(art, node48->children[idx]);
+                }
+            }
+        } break;
+        case CROARING_ART_NODE256_TYPE: {
+            art_node256_t *node256 = (art_node256_t *)art_deref(art, ref);
+            for (int i = 0; i < 256; ++i) {
+                if (node256->children[i] != CROARING_ART_NULL_REF) {
+                    node256->children[i] =
+                        art_move_node_to_shrink(art, node256->children[i]);
+                    art_shrink_at(art, node256->children[i]);
+                }
+            }
+        } break;
+        default:
+            assert(false);
+            break;
+    }
 }
 
-art_val_t *art_erase(art_t *art, const art_key_chunk_t *key) {
-    if (art->root == NULL) {
-        return NULL;
+static bool art_is_shrunken(const art_t *art) {
+    return art->first_free[CROARING_ART_LEAF_TYPE] ==
+               art->capacities[CROARING_ART_LEAF_TYPE] &&
+           art->first_free[CROARING_ART_NODE4_TYPE] ==
+               art->capacities[CROARING_ART_NODE4_TYPE] &&
+           art->first_free[CROARING_ART_NODE16_TYPE] ==
+               art->capacities[CROARING_ART_NODE16_TYPE] &&
+           art->first_free[CROARING_ART_NODE48_TYPE] ==
+               art->capacities[CROARING_ART_NODE48_TYPE] &&
+           art->first_free[CROARING_ART_NODE256_TYPE] ==
+               art->capacities[CROARING_ART_NODE256_TYPE];
+}
+
+void art_init_cleared(art_t *art) {
+    art->root = CROARING_ART_NULL_REF;
+    memset(art->first_free, 0, sizeof(art->first_free));
+    memset(art->capacities, 0, sizeof(art->capacities));
+    art->leaves = NULL;
+    art->node4s = NULL;
+    art->node16s = NULL;
+    art->node48s = NULL;
+    art->node256s = NULL;
+}
+
+size_t art_shrink_to_fit(art_t *art) {
+    if (art_is_shrunken(art)) {
+        return 0;
     }
-    art_erase_result_t result = art_erase_at(art->root, key, 0);
-    if (result.value_erased == NULL) {
-        return NULL;
+    if (art->root != CROARING_ART_NULL_REF) {
+        art->root = art_move_node_to_shrink(art, art->root);
+        art_shrink_at(art, art->root);
+    }
+    return art_shrink_node_arrays(art);
```

Comment:
I don't think this works. I need to think more to verify, but imagine the case where we have:

Capacity: 8
Root: idx 5
Free List: 6 -> 4 -> 8

The root will not be moved (because it's before `first_free`). At some point, the item at idx 7 will be moved into slot 6. But the item at 5 cannot be moved into free slot 4, because we already visited the root.

Reply:
I changed this to sort the free list during shrinking to avoid this issue. Not sure how to test though, suggestions welcome.